packages/minter-contracts/test-hs/Test/EnglishAuction/TezFixedFee.hs

```diff
@@ -0,0 +1,696 @@
+module Test.EnglishAuction.TezFixedFee where
+
+import Cleveland.Util (sec)
+import qualified Data.List as List
+import qualified Data.Map as Map
+import Data.Traversable.WithIndex (TraversableWithIndex, ifor)
+import Fmt ((+|), (|+))
+import Hedgehog (Gen, Property, forAll, label, property)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Indigo.Contracts.FA2Sample as FA2
+import Lorentz hiding (amount, contract, now)
+import Lorentz.Contracts.EnglishAuction.TezFixedFee
+import Lorentz.Contracts.MinterSdk
+import Lorentz.Contracts.Spec.FA2Interface (TokenId(..))
+import Michelson.Typed (convertContract, untypeValue)
+import Morley.Nettest
+import Test.Util (balanceOf, clevelandProp, genMutez')
+import Tezos.Core (timestampPlusSeconds)
+
+hprop_Auctions_must_start_within_the_configured_number_of_seconds :: Property
+hprop_Auctions_must_start_within_the_configured_number_of_seconds =
+  property $ do
+    testData@TestData{testMaxConfigToStartTime} <- forAll genTestData
+
+    testTimeToStart <- forAll $ Gen.integral $
+      Range.linear
+        (toInteger testMaxConfigToStartTime + 1)
+        ((toInteger testMaxConfigToStartTime + 1 ) * 2)
+    let testData' = testData { testTimeToStart = testTimeToStart }
+
+    clevelandProp $ do
+      setup@Setup{contract} <- testSetup testData'
+
+      configureAuction testData' setup
+        `expectFailure`
+          failedWith contract [mt|start_time must not be greater than the sum of current time and max_config_to_start_time|]
+
+hprop_First_bid_is_valid_IFF_it_meets_opening_price :: Property
+hprop_First_bid_is_valid_IFF_it_meets_opening_price =
+  property $ do
+
+    testData@TestData{testOpeningPrice} <- forAll genTestData
+    firstBid <- forAll $ genMutez' (Range.linear 0 (testOpeningPrice * 2))
+
+    clevelandProp $ do
+      setup@Setup{seller, startTime, contract} <- testSetup testData
+      bidder <- newAddress "bidder"
+      withSender seller $ configureAuction testData setup
+      waitForAuctionToStart testData
+
+      withSender bidder $ do
+        if firstBid >= testOpeningPrice
+          then placeBid contract firstBid
+          else placeBid contract firstBid `expectFailure` failedWith contract
+                 ([mt|Invalid Bid amount|] :: MText, ((testOpeningPrice, firstBid) , seller, startTime), startTime)
```

Comment:
Hmm I tried changing our code to this:

```
([mt|Invalid Bid amount|], (testOpeningPrice, firstBid, seller, startTime, startTime))
```

Which seems to match the corresponding ligo code:

```
("Invalid Bid amount", (auction.current_bid, Tezos.amount, auction.highest_bidder, auction.last_bid_time, Tezos.now)) : unit)
```

This fails though:

```
              Expected exception due to:
                Contract: KT18zy4cNdbpT8eJvAdVgP7e4AVegBdvztTS failed with: (Invalid Bid amount, Pair (100, 0) , 12437Vg67WNe9bXpvA4ZVVdPEg7U7a5hXVU, 100, 100)
              But got:
                Contract: KT18zy4cNdbpT8eJvAdVgP7e4AVegBdvztTS failed with: (Invalid Bid amount, (Pair (100, 0) , 12437Vg67WNe9bXpvA4ZVVdPEg7U7a5hXVU, 100), 100)
```

Maybe there's a difference in how ligo and lorentz treat tuples with 3 or more members? The [ligo spec](https://ligolang.org/docs/language-basics/sets-lists-tuples) doesn't say anything about this....

Reply:
That's quite odd, looking at the result and the ligo code it looks like the pair is handled as left-comb.

Looking at this other [spec page](https://ligolang.org/docs/advanced/interop) it is a bit unclear how tuples are handled, but if we assume that they are like records, then [this](https://ligolang.org/docs/advanced/interop#default-ligo-output) applies:
> By default LIGO translates its datatypes into a alphabetically left balanced tree.

which (aside from the sorting, which doesn't happen since tuple fields aren't named) seems to be correct.

Lorentz indeed doesn't do this, AFAIK there is no advantage in using left-combed trees for tuples.

Reply:
Or sorry, I missed this comment somehow.

Yeah, we handle tuples in the same way how Haskell Generics work (right-balanced tree?), and that error in LIGO is represented as `pair (pair · ·) (pair · ·) ·` - looks like this is just a coincidence that their representations matched ))

Mmm, actually I believed that types like `a * b * c ...` translate to just comb tree, but actually, it translates to LIGO's default layout, and handling that manually in Haskell with just pairs is inconvenient. Probably the only suitable way would be to create a dedicated datatype for this error :thinking: